src/common/entity/compute_state_display.ts

```diff
@@ -29,37 +29,77 @@ export const computeStateDisplay = (
   const domain = computeStateDomain(stateObj);
 
   if (domain === "input_datetime") {
-    let date: Date;
-    if (!stateObj.attributes.has_time) {
+    if (state) {
+      // If trying to display an explicit state, need to parse the explict state to `Date` then format.
+      // Attributes aren't available, we have to use `state`.
+      try {
+        // We are going to convert the state string into a ISO8601 string,
+        // which is always in UTC and causing the tiemzone info to be lost,
+        // therefore we need to compensate the local timezone offset.
+        const timezoneOffset = new Date().getTimezoneOffset() * 60 * 1000;
+        const components = state.split(" ");
+        if (components.length === 2) {
+          // Date and time.
+          const iso8601String = components.join("T") + "Z";
+          const dateObj = new Date(iso8601String);
+          const adjustedDateObj = new Date(dateObj.getTime() + timezoneOffset);
+          return formatDateTime(adjustedDateObj, locale);
+        }
+        if (components.length === 1) {
+          if (state.includes("-")) {
+            // Date only.
+            // Date only state is already a valid ISO8601 string.
+            const dateObj = new Date(state);
+            const adjustedDateObj = new Date(
+              dateObj.getTime() + timezoneOffset
+            );
+            return formatDate(adjustedDateObj, locale);
```

Comment:
This will work, but `const dateObj = new Date(`${state}T00:00:00`);` will work too.